### PR TITLE
PPC64 part of JDK-8281146

### PR DESCRIPTION
**Pull request by TheRealMDoerr**
It can get implemented similarly on PPC64.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Error
&nbsp;⚠️ This PR only contains changes already present in the target

### Integration blocker
&nbsp;⚠️ Dependency #7231 must be integrated first

### Reviewers
 * [Richard Reingruber](https://openjdk.java.net/census#rrich) (@reinrich - **Reviewer**) ⚠️ Review applies to dd124027db06326232f565c09ef3450a87d996fa
 * [Lutz Schmidt](https://openjdk.java.net/census#lucy) (@RealLucy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7430/head:pull/7430` \
`$ git checkout pull/7430`

Update a local copy of the PR: \
`$ git checkout pull/7430` \
`$ git pull https://git.openjdk.java.net/jdk pull/7430/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7430`

View PR using the GUI difftool: \
`$ git pr show -t 7430`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7430.diff">https://git.openjdk.java.net/jdk/pull/7430.diff</a>

</details>

### CODE REVIEW

**Comment by TheRealMDoerr** [comment]
Hi Richard,
thanks for reviewing it!
The match rules are optional. (`Matcher::match_rule_supported(Op_HasNegatives)` checks if ALDC has created one.) Note that the same replacement was done for x86 and some other platforms don't have it at all (arm32).
Yes, I have used our nightly tests to test Claes' change together with my PPC64 implementation.
I guess it will be better to let Claes handle Copyright year updates when all changes are complete. Otherwise, merge conflicts may arise.

**Comment by reinrich** [comment]
> Hi Richard, thanks for reviewing it! The match rules are optional. (`Matcher::match_rule_supported(Op_HasNegatives)` checks if ALDC has created one.)

Thanks for the indication. I expected that but couldn't find it.

> Note that the same replacement was done for x86 and some other platforms don't have it at all (arm32). Yes, I have used our nightly tests to test Claes' change together with my PPC64 implementation.

Good.

> I guess it will be better to let Claes handle Copyright year updates when all changes are complete. Otherwise, merge conflicts may arise.

Ok.

Cheers, Richard.

**Comment by cl4es** [comment]
Thanks so much for doing this, Martin!

I aim to wait for aarch64 and s390 variants to be implemented as well and remove all the code for supporting the `_hasNegatives` intrinsic in the hotspot code.

**Comment by backwaterred** [comment]
I am taking a look at this PR to better understand how intrinsics function in Hotspot, and I noticed that the old intrinsic name is mentioned [here](https://github.com/openjdk/jdk/pull/7430/files#diff-d21a64a4949f298476bf91083d3b956face9a6393a08a706b071068898533082R12788).

Nice changes! They are clean and easy to understand.

File: src/hotspot/cpu/ppc/ppc.ad [hunk 1]
```diff
@@ -12779,16 +12779,16 @@ instruct string_inflate(Universe dummy, rarg1RegP src, rarg2RegP dst, iRegIsrc l
...
+instruct count_positives(iRegPsrc ary1, iRegIsrc len, iRegIdst result, iRegLdst tmp1, iRegLdst tmp2,
+                         regCTR ctr, flagsRegCR0 cr0)
 %{
-  match(Set result (HasNegatives ary1 len));
-  effect(TEMP_DEF result, USE_KILL ary1, TEMP tmp1, TEMP tmp2, KILL ctr, KILL cr0);
+  match(Set result (CountPositives ary1 len));
+  effect(TEMP_DEF result, TEMP tmp1, TEMP tmp2, KILL ctr, KILL cr0);
   ins_cost(300);
   format %{ "has negatives byte[] $ary1,$len -> $result \t// KILL $tmp1, $tmp2" %}
   ins_encode %{
```
Comment (RealLucy) [inline src/hotspot/cpu/ppc/ppc.ad:12789]: How about adapting the cleartext name in the format? I would suggest to use the same name as on x86 (countPositives).

**Review by RealLucy — CHANGES_REQUESTED** [review]

**Comment by cl4es** [comment]
Pre-existing, but I note that the ppc (and s390) impl don't have an explicit effect on the `len` register. x86 and aarch64 does a `USE_KILL len`. Not exactly sure what this means exactly, or whether it's intentional or not. What do you think?

**Comment by RealLucy** [comment]
USE_KILL reg means the reg contents on entry to the code is relevant (in contrast to TEMP), and the reg contents is altered on exit from the code. 

ppc and s390 don't alter the len reg. There is no reason to prevent the compiler from re-using the reg contents afterwards.

**Review by RealLucy — APPROVED** [review]
Changes look good to me.

**Comment by TheRealMDoerr** [comment]
Integrated by https://github.com/openjdk/jdk/pull/7231/commits/b826ef8102cddf0afc22d9d56111d5fba9d1e091

**Pull request closed without merging**